lib/Entity/Folder.php

```diff
@@ -0,0 +1,344 @@
+
+        if ($mode === 'delete') {
+
+            // remove this folder from children of the parent
+            foreach ($parentChildren as $index => $child) {
+                if ((int)$child === (int)$this->id) {
+                    unset($parentChildren[$index]);
+                }
+            }
+
+            // remove this folder children
+            foreach ($children as $child) {
+                $childObject = $this->folderFactory->getById($child);
+                $childObject->manageChildren('delete');
+                $this->getStore()->update('DELETE FROM `folder` WHERE folderId = :folderId', [
+                    'folderId' => $childObject->id
+                ]);
+            }
+        } else {
+            $parentChildren[] = $this->id;
+        }
+
+        $updatedChildren = implode(',', array_filter($parentChildren));
+
+        $this->getStore()->update('UPDATE `folder` SET children = :children WHERE folderId = :folderId', [
+                'folderId' => $this->parentId,
+                'children' => $updatedChildren
+            ]);
+    }
+
+    public function managePermissions()
+    {
+        // this function happens after permissions are inserted into permission table
+        // with that we can look up if there are any permissions for edited folder and act accordingly.
+        $permissionExists = $this->getStore()->exists('SELECT permissionId FROM permission INNER JOIN permissionentity ON permission.entityId = permissionentity.entityId WHERE objectId = :folderId AND permissionentity.entity = :folderEntity', [
+            'folderId' => $this->id,
+            'folderEntity' => 'Xibo\Entity\Folder'
+        ]);
+
+        if ($permissionExists) {
+            // if we added/edited permission on this folder, then new ACL starts here, cascade this folderId as permissionFolderId to all children
+            $this->getStore()->update('UPDATE `folder` SET permissionsFolderId = NULL WHERE folderId = :folderId', [
+                'folderId' => $this->id
+            ]);
+            $permissionFolderId = $this->id;
+        } else {
+            // if there are no permissions for this folder, basically reset the permissions on this folder and its children
+            if ($this->id === 1 && $this->isRoot()) {
+                $permissionFolderId = 1;
+            } else {
+                $parent = $this->folderFactory->getById($this->parentId);
+                $permissionFolderId = ($parent->permissionsFolderId == null) ? $parent->id : $parent->permissionsFolderId;
+            }
+
+            $this->getStore()->update('UPDATE `folder` SET permissionsFolderId = :permissionFolderId WHERE folderId = :folderId', [
+                'folderId' => $this->id,
+                'permissionFolderId' => $permissionFolderId
+            ]);
+        }
+
+        $this->getStore()->update('UPDATE `folder` SET permissionsFolderId = :permissionsFolderId WHERE parentId = :folderId',[
```

Comment:
We could extract a method here, so that you didn't have to repeat the same below for the children?